lib/auth/server_monitor.go

```diff
@@ -0,0 +1,161 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	notificationsv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/notifications/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/inventory"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+const (
+	// systemClockCheckCycle is the period when system clock comparison is launched
+	// across all inventories, to be gathered for global notifications, if any.
+	systemClockCheckCycle = 10 * time.Minute
+	// systemClockThreshold is the duration threshold for triggering a warning
+	// if the time difference exceeds this threshold.
+	systemClockThreshold = time.Minute
+	// systemClockNotificationWarningName is the ID for adding the global notification.
+	systemClockNotificationWarningName = "cluster-monitor-system-clock-warning"
+	// systemClockMessagesLimit is limit for showing the list of affected inventories.
+	systemClockMessagesLimit = 10
+)
+
+// MonitorSystemTime runs the periodic check for iterating through all inventories
+// to ping them and receive the system clock difference.
+func (a *Server) MonitorSystemTime(ctx context.Context) error {
+	checkInterval := interval.New(interval.Config{
+		FirstDuration: time.Second * 10,
+		Duration:      systemClockCheckCycle,
+		Clock:         a.GetClock(),
+	})
+	defer checkInterval.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-checkInterval.Next():
+			a.checkInventorySystemClocks(ctx)
+		}
+	}
+}
+
+// checkInventoryClocks iterates through inventory store instance state to gather
+// information about the system clock differences.
+func (a *Server) checkInventorySystemClocks(ctx context.Context) {
+	var counter int
+	var messages []string
+	a.inventory.Iter(func(handle inventory.UpstreamHandle) {
+		counter++
+		if counter >= systemClockMessagesLimit {
+			return
+		}
+
+		hello := handle.Hello()
+		handle.VisitInstanceState(func(ref inventory.InstanceStateRef) (update inventory.InstanceStateUpdate) {
+			if ref.LastHeartbeat != nil && ref.LastHeartbeat.GetLastMeasurement() != nil {
+				m := ref.LastHeartbeat.GetLastMeasurement()
+				diff := m.ControllerSystemClock.Sub(m.SystemClock) - m.RequestDuration/2
```

Comment:
This might deserve a little comment explaining why we're subtracting half of the RTT.